### PR TITLE
[source-google-ads]: Concurrent CDK and performance fixes

### PR DESCRIPTION
**Pull request by nurikk**
This PR introduces following changes:

1) Refactors `SourceGoogleAds` to use Concurrent CDK #36852

### CODE REVIEW

**Comment by marcosmarxm** [comment]
@natikgadzhi can extensibility help reviewing this change?

**Comment by natikgadzhi** [comment]
Sure. @nurikk, thank you for putting this together! Concurrent CDK is new, and we're careful about it. And google analytics is a popular source, so we want to get this right.

Ideally, I would love to get a review from @lazebnyi, @clnoll, and maybe @bazarnov? Let's use the live testing tool on this and see if concurrent approach works well, is fast, and doesn't drop records.

**Comment by clnoll** [comment]
@nurikk thanks for submitting this! In your PR description, should "2. Fixes connection validation loop" link to a different issue? I want to make sure I have the full context.

File: airbyte-integrations/connectors/source-google-ads/unit_tests/test_source.py
Comment (clnoll) [inline airbyte-integrations/connectors/source-google-ads/unit_tests/test_source.py:429]: Where does the 5 come from? Mind adding a code comment for it?

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/spec.json
Comment (clnoll) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/spec.json:79]: WDYT about doing this in a separate PR so we can minimize the changes that we introduce when adding concurrency?

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py
Comment (clnoll) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py:81]: Since we do `concurrency_level // 2` here, the min concurrency in the config needs to be 2.

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py
Comment (clnoll) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py:153]: Would you mind updating the description in spec.json to state that if no manager ID is present we'll get all customers for all accounts? (This can be done in the separate PR if we move the manager ID logic there.)

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py
Comment (clnoll) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py:235]: Is this just a performance optimization? Can you update the comment to explain why we only need one of each for validation?

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/state_converter.py
Comment (clnoll) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/state_converter.py:16]: Is there some documentation that we can link that indicates that requests should be made for whole days?

**Review by clnoll — CHANGES_REQUESTED** [review]
@nurikk this looks great! My main question is whether we can split out some of the logic into separate PRs. Since adding concurrency is a pretty big change, I'd like us to release it on its own.

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py
Comment (nurikk) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py:81]: Fixed

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/state_converter.py
Comment (nurikk) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/state_converter.py:13]: I don't know where I can find proof for this. But the lowest granularity for google ads api is day

**Comment by nurikk** [comment]
Hi @clnoll, thanks for review! I've addressed all your comments. Now this PR is only about concurrent CDK

**Comment by natikgadzhi** [comment]
@katmarkham this seems like a really good work, but we're out of capacity to push this over the finish line. It's prob up to you and @maxi297 to assign to and complete.

**Comment by maxi297** [comment]
Here are some thoughts on the current situation.

## Summary

Not all streams can be easily moved to concurrent.

### IncrementalGoogleAdsStream

I think nurikk did a pretty good job of migrating this to concurrency. For now, I'll highlight some points. I can QA a bit more tomorrow to identify edge cases.

#### State by parent id
We currently don't support state by parent. In this case, the parents are the customers. This was addressed by nurikk [here](https://github.com/airbytehq/airbyte/pull/36853/files#diff-a13653278aaa4b9fc1ae5e423df0580257c7890716bfea85545c3cbcb41e8638) and [here](https://github.com/airbytehq/airbyte/pull/36853/files#diff-0e2625161bb381226e2f6b158f2eb98b1a27885608225605f01a109ba177c54e). We are moving to a model where the state converter should do very little (just the conversion of the types from state to cursor value and vice-versa) but for now we should be fine with this implementation. 

#### Checkpoint interval
This feature is not supported in the Concurrent CDK but I'm fine with removing this to support concurrency. This would be better if we were to have partitioned state but this would be a breaking change though. This hasn't been addressed as part of this PR but I wouldn't make this a blocker. We can evaluate empirically in prod if this is generating too much flakiness.



### IncrementalEventsStream

#### Slices are not json serializable

The slices contain sets for `updated_ids` and `deleted_ids` which are not serializable.

We can either:
* use a custom encoder to support sets in slices in the CDK:
```
class SetEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, set):
            return list(obj)
        return json.JSONEncoder.default(self, obj)
```
... and use it in `StreamPartition.__hash__`
* change the slices to have lists instead of sets

#### State multiple formats

Ultimately, I think this is the point that makes if very hard to migrate to concurrency. I don't have a good grasp of the possible state format but based on [this](https://github.com/airbytehq/airbyte/blob/0766bd05459a6d66bffd923fd8109cc936e75fdb/airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py#L750-L755), the state can have two levels of parent i.e. parent stream name and customer_id or be flat if the state is not by customer id. I don't have enough knowledge to know if both are actually used and if it is the case, explain why. 


## Proposition

### To unblock this PR

#### Remove IncrementalEventsStreams from concurrency

Those streams are a bit more complex and I don't have a good grasp of the state to propose a solution just yet. It shouldn't prevent us from adding `IncrementalGoogleAdsStream` as concurrent streams which nurikk would benefit from immediately for `IncrementalCustomQuery`.

#### Update to the newest CDK version

It is important to update the version of the airbyte_cdk because if there is an issue in prod that requires us to update the CDK we would be blocked as there were breaking changes since the last release ; we added the slice generation logic in the ConcurrentCursor and pushed some logic from the state adapters to the cursor. For this specific PR, it means providing `end_provider` to the cursor even though it won't be used yet and rename the method `convert_to_sequential_state` to `convert_to_state_message`. We don't need to block the PR for not using the slice generation from the Concurrent CDK.



### Long term

#### Generalize the logic

In order for other sources to leverage this, we should add the parent cursor mechanism to the Concurrent CDK. This is a very common pattern that will block the migration of multiple connectors for concurrency.


#### Improve resiliency

The newest CDK version allows for generating partitioned state which would compensate for the removal of the checkpoint interval. The change should be straighforward and we should be able to do this by enabling partitioned state using the feature flag

This will be a breaking change as the format of the state will be different. However, this change is backward compatible so we would not need the opt-in mechanism.

Poking @girarda to see if this makes sense on the Extensibility and API Sources side.

**Comment by girarda** [comment]
@maxi297 I agree with everything you said except I think it should be possible to migrate to partitioned state without a breaking change - although we might still want to introduce a breaking change to stop supporting the multiple state formats you mention. This can be done separately from this PR so no need to block on this.

**Comment by maxi297** [comment]
Adding notes from our grooming:
* There might be some risk about having a low rate limit for Google Ads. We should make sure we support users that have different rate limits
    * @maxi297 to validate if we hit the rate limit today
    * @maxi297 to validate client side rate limiting for gRPC/Google Ads
* Just to clarify: the breaking change is not on the source level. We just need to update the CDK so that devs don't have to deal with surprise changes

**Comment by nurikk** [comment]
Hi @maxi297 ! I've made required changes to unblock this PR. Please help me with review

**Pull request closed without merging**
